syntax = "proto3";

package trafficgen;

// The main service definition.
service TrafficGeneratorService {
  // Configure the IMIX (Internet Mix) profile for traffic generation.
  rpc ConfigureIMIX(IMIXConfig) returns (StatusResponse) {}

  // Configure a set of specific flows.
  rpc ConfigureFlows(FlowConfig) returns (StatusResponse) {}

  // Start traffic generation for all configured flows.
  rpc StartTraffic(StartRequest) returns (StatusResponse) {}

  // Stop traffic generation.
  rpc StopTraffic(StopRequest) returns (StatusResponse) {}

  // Get real-time metrics.
  rpc GetMetrics(MetricsRequest) returns (MetricsResponse) {}

  // Get summary statistics.
  rpc GetStats(StatsRequest) returns (StatsResponse) {}
}

// --- Common Messages ---

// Generic status response for configuration RPCs.
message StatusResponse {
  bool success = 1;
  string message = 2;
  int32 error_code = 3;
}

// --- IMIX Configuration ---

// An entry in the IMIX profile.
message IMIXEntry {
  uint32 packet_size = 1;
  float percentage = 2;
  string protocol = 3;
  bool enable_checksum = 4;
}

// Configuration for an IMIX traffic profile.
message IMIXConfig {
  repeated IMIXEntry entries = 1;
  uint64 target_pps = 2;
  uint32 duration_seconds = 3; // 0 for infinite
}

// --- Flow Configuration ---

// A single traffic flow definition.
message Flow {
  uint32 flow_id = 1;
  string src_ip = 2;
  string dst_ip = 3;
  uint32 src_port = 4;
  uint32 dst_port = 5;
  string protocol = 6; // "tcp", "udp", "icmp"
  uint32 packet_size = 7; // If 0, IMIX profile is used
  uint64 pps = 8;
  uint32 duration_seconds = 9; // 0 for infinite
  string dst_mac = 10; // Destination MAC address
}

// A request to configure multiple flows.
message FlowConfig {
  repeated Flow flows = 1;
  bool stateless = 2; // True for stateless, False for stateful (e.g., TCP)
}

// --- Traffic Control ---

message StartRequest {
  bool reset_stats = 1;
}

message StopRequest {
  bool drain_queues = 1;
  int32 timeout_seconds = 2;
}

// --- Metrics and Statistics ---

message MetricsRequest {
  bool include_per_core = 1;
  bool include_latency_histogram = 2;
}

message GlobalMetrics {
  uint64 total_tx_packets = 1;
  uint64 total_rx_packets = 2;
  uint64 total_tx_bytes = 3;
  uint64 total_rx_bytes = 4;
  double current_pps = 5;
  double current_bps = 6;
  double avg_latency_us = 7;
  uint64 errors = 8;
}

message CoreMetrics {
  uint32 core_id = 1;
  uint64 tx_packets = 2;
  uint64 rx_packets = 3;
  uint64 tx_bytes = 4;
  uint64 rx_bytes = 5;
  double pps = 6;
}

message LatencyBucket {
    uint32 min_us = 1;
    uint32 max_us = 2;
    uint64 count = 3;
}

message LatencyHistogram {
    repeated LatencyBucket buckets = 1;
}

message MetricsResponse {
  GlobalMetrics global = 1;
  repeated CoreMetrics per_core = 2;
  LatencyHistogram latency = 3;
}

message StatsRequest {
  bool detailed = 1; // Include per-flow stats
}

message FlowStats {
  uint32 flow_id = 1;
  uint64 tx_packets = 2;
  uint64 tx_bytes = 3;
  double avg_pps = 4;
}

message StatsResponse {
  GlobalMetrics metrics = 1;
  uint64 uptime_seconds = 2;
  map<uint32, FlowStats> flow_stats = 3;
}
